include/os/linux/kernel/linux/simd_x86.h

```diff
@@ -126,96 +124,191 @@
 #endif
 
 #else /* defined(KERNEL_EXPORTS_X86_FPU) */
+
 /*
  * When the kernel_fpu_* symbols are unavailable then provide our own
- * versions which allow the FPU to be safely used in kernel threads.
- * In practice, this is not a significant restriction for ZFS since the
- * vast majority of SIMD operations are performed by the IO pipeline.
+ * versions which allow the FPU to be safely used.
  */
+#if defined(HAVE_KERNEL_FPU_INTERNAL)
+
+extern struct fpu **zfs_kfpu_fpregs;
 
 /*
- * Returns non-zero if FPU operations are allowed in the current context.
+ * Initialize per-cpu variables to store FPU state.
  */
-#if defined(HAVE_KERNEL_TIF_NEED_FPU_LOAD)
-#define	kfpu_allowed()		((current->flags & PF_KTHREAD) && \
-				test_thread_flag(TIF_NEED_FPU_LOAD))
-#elif defined(HAVE_KERNEL_FPU_INITIALIZED)
-#define	kfpu_allowed()		((current->flags & PF_KTHREAD) && \
-				current->thread.fpu.initialized)
-#else
-#define	kfpu_allowed()		0
-#endif
+static inline void
+kfpu_fini(void)
+{
+	int cpu;
+
+	for_each_possible_cpu(cpu) {
+		if (zfs_kfpu_fpregs[cpu] != NULL) {
+			kfree(zfs_kfpu_fpregs[cpu]);
+		}
+	}
+
+	kfree(zfs_kfpu_fpregs);
+}
+
+static inline int
+kfpu_init(void)
+{
+	int cpu;
+
+	zfs_kfpu_fpregs = kzalloc(num_possible_cpus() *
+	    sizeof (struct fpu *), GFP_KERNEL);
+	if (zfs_kfpu_fpregs == NULL)
+		return (ENOMEM);
+
+	for_each_possible_cpu(cpu) {
+		zfs_kfpu_fpregs[cpu] = kmalloc_node(sizeof (struct fpu),
+		    GFP_KERNEL, cpu_to_node(cpu));
```

Comment:
Good thought.  I've updated the patch to zero this memory when initially allocated.  As for the need to re-initialize after a restore since we're using dedicated memory for this only available to zfs I don't believe we need to re-initialize it after each restore.

Reply:
sorry for not being verbose enough - what I am worried about is

task A does something with FPU
ZFS saves state (does this always write the full fsave/fxsave/xsave struct?)
ZFS restores state
task A
kernel clears HW FPU state if needed
task B
ZFS saves state (same question?)
ZFS restore (if the two questions above are not a yes, then this might now restore partial state from task A)
task B

or

task A does something with FPU
ZFS saves state (same question)
ZFS restores state
ZFS saves state (same question)
ZFS restores state
task A (potentially sees (partial) state from ZFS FPU operations in HW registers)

it's probably just a question of reading the spec - I won't have time for that today unfortunately, but hopefully on Monday.